### DISCUSSION TITLE
PHP 8x support

### DISCUSSION THREAD

**Discussion by NathanGibbs3** [NathanGibbs3/BASE#171] [Ideas]
### Open discussion about the evolving state of PHP 8x support in BASE.
---
The idea is to try to keep the discussion about this in one place and not have fragments of it splattered chaotically all across the issue tracker. 
 - Lets use the issue tracker to track issues.
 - Lets use this topic to have & track ideas.

**Comment by mesteele** [comment]
Navigating to the Use Archive Database all the items under Traffic Profile by Protocol is missing until an event is moved into it and then they all display. If I delete the event, they all disappear.
![2023-04-08_10-35-14](https://user-images.githubusercontent.com/12111398/230727367-17233245-7dcb-4513-b6cb-105b0c2ff410.png)

**Reply by NathanGibbs3** [reply]
Can you test the following.
 - Does BASE do this with an empty ALERT DB as well ?
 - Does BASE also do this on PHP 7x ?

**Reply by mesteele** [reply]
PHP 7 works fine with and without events in either the Alert or the Archive database.

Another unrelated issue to the above: The protocol list should be showing 0% when it is empty, and it shows NAN%. If one event is received all the protocols lose the NAN and show correctly. This has been doing the same on PHP7.

PHP 8: I cleared all the events, and it loses all the protocols as it shows above in both the Alert and Archive database.

**Reply by mesteele** [reply]
I just restarted at zero events and all the protocols were missing on the Alert and Archive database. As soon as it received one event in the Alert database all the protocols showed up and the percentage shows as <1% on the ones with no events, and the one with 1 event shows as 100%

**Reply by NathanGibbs3** [reply]
I think I found the source of the NAN Issue last night while reviewing the code for this issue, committ 5d91e74ae7f6 was responsible for that. 
Somebody, two guesses who, was writing code without enough coffee. :smile:
I already have a fix for that in the pipeline.

Lets make sure I understand this issue correctly.
 - We are getting the bar graphs at 0% on PHP 7x.
 - We are getting a blank section of page on PHP 8s.

Do I have the right idea?

**Reply by mesteele** [reply]
If the database is new there are no events posted. The bar graphs on PHP8 are missing. As soon as it receives one event in one of the protocols all the graphs show up. The graph that got the event will show 100% and the remaining graphs will show <1%

On PHP7 all the bar graphs are visible with no events logged and all the percentage values are NAN%. As soon as it receives one event in one of the protocols the graph that got the event will show 100% and the remaining graphs will show <1%

**Reply by NathanGibbs3** [reply]
OK, got it.
The next commit to devel will have the NAN fix, I suspect that will also fix the PHP 8x display issue, if not we can go back to the drawing board. Either way, I opened Issue #172 to track this.
I'm going to do some work on #161 before pushing to devel, so it will be a bit.

**Comment by mesteele** [comment]
If $Use_Auth_System = 0; I can still get into Alert Group Maintenance and create or delete.
![2023-04-08_10-38-23](https://user-images.githubusercontent.com/12111398/230727484-1558bc7a-ef1c-4276-a520-7b0aa6423be7.png)

**Reply by NathanGibbs3** [reply]
Not specifically related to PHP 8x support.
Issue #161 is open to track this.
If the Auth system is off, I don't see any issues with allowing users full access to alert group functions.
If the Auth system is on, there is the expectation of access levels.
The create, edit, and delete functions should be locked out for users with in roles with numeric ids above ag_editor, currently, they are not.

Maybe the clear function could be locked out as well, your thoughts on that would be appreciated @mesteele .

I am not entirely sure what usage scenarios were envisioned by @secureideas when they built the auth system.
One thing that is clear is that the auth system was never fully implemented,  although we have made progress with that.🙂

Added a new project to track Auth System related Issues https://github.com/NathanGibbs3/BASE/projects/5

**Reply by mesteele** [reply]
It's a lousy authentication system. I'm not sure how it should be, but I don't think users should be able to modify.

**Reply by NathanGibbs3** [reply]
Agreed, the auth system tables should not even be part of the snort DB.
They should have been implemented as a separate DB, like the Archive DB.
On the bright side, your observations have led the creation of Issues such as #144 & #146, which make the auth system suck less.

**Comment by mesteele** [comment]
Most of the links in the home page produce an error when accessed and they are directed at that particular database table related to the link.
![2023-04-08_10-37-20](https://user-images.githubusercontent.com/12111398/230727620-2de9630b-bc36-4b5e-bfea-15d443bf3d84.png)

**Reply by NathanGibbs3** [reply]
- Is the latest commit from the devel branch doing this ?
 - Is the same commit, connecting to the same DB, running OK on PHP 7x ?

**Reply by mesteele** [reply]
I'm grabbing the DEV branch, correct? I just grabbed a new one just before testing.

All the links work in PHP7

**Reply by NathanGibbs3** [reply]
Correct, the devel branch, you are on the latest commit, I added the text 'Query Fail:' to the error message in the latest commit last night, so disregard the first question.

It would appear that something is different in PHP 8, as the DB query isn't even getting a Recordset.
Send error logs.

**Reply by mesteele** [reply]
Apache log:
[error.log](https://github.com/NathanGibbs3/BASE/files/11184214/error.log)

**Reply by NathanGibbs3** [reply]
Thank you, that is helpful.
The SQL in the error log is definitely broken.
I tested some of it on my DB's cli and had similar results.

This tells us that the SQL generated by BASE running on
 - PHP 8x is invalid.
 - PHP 7x is valid.

I think the root cause is issue #168.
For fun, check if the Unique IP links page works from you BASE main page on PHP 8x.
Based on the URLs in your screenshots, this should take you directly to the page on your setup, if the main page is still not loading correctly for you under PHP 8x. 
http://winids/base_stat_iplink.php

I am wondering if the fixes for #168 on that page fixed the invalid SQL issues there or not.

Another thing to consider, although unlikely to be the culprit, is ADOdb.
Specifically if you were running different versions between PHP 7x & PHP 8x.
However if you are running the same ADOdb version on PHP 7x & PHP 8x, I doubt that would be the problem.

**Reply by mesteele** [reply]
ADOdb is the same for PHP7 and PHP8

The http://winids/base_stat_iplink.php logs to a blank page.
![2023-04-09_19-25-48](https://user-images.githubusercontent.com/12111398/230801462-c06106f1-164b-4964-b7f0-4a3bf27c08e9.png)

The other links are not showing an error now, but the links displays this on them all.
![2023-04-09_19-28-04](https://user-images.githubusercontent.com/12111398/230801493-0362d8e6-f5b2-4933-be5c-66ec9e610bf0.png)

Log:
[error.log](https://github.com/NathanGibbs3/BASE/files/11186533/error.log)

**Reply by NathanGibbs3** [reply]
Awesome, so we rule out different ADOdb versions as the culprit. 👍 

The codebase that generated the above log file cannot possibly be the latest from the devel branch.
It has issue #166 which was closed by commit c1b330c days ago.
What you gave me was a log file full of errors that are already fixed in the devel branch.

All the links that don't work lead to files with issue #168.
I am currently working on fixing that in all of them.

However, the codebase from commit 805087 already has the issue #168 fix for base_stat_iplink.php
Please, test that commit on PHP 8x, specifically the unique IP Links page. base_stat_iplink.php

What I am hoping to learn is if there are more issues that will pop up once #168 is cleared away.
We can wait until tomorrow or the next day, when I eliminate the issue from all files & push to devel, or we can test the one file in the repo that already has the issue fixed and find out sooner.

**Reply by mesteele** [reply]
Whoops, I grabbed the Master. Pulling the correct one now :)

Ok, when a new push is completed I'll give it a go.

**Reply by NathanGibbs3** [reply]
I pushed to devel earlier, CI checks passed, let me know if that commit is better behaved on PHP 8x.
Knowing PHP 8x, something else will probably blow up. 😄
Fingers crossed. 🤞 😄

**Reply by mesteele** [reply]
Most of the links on the home page are still broken, similar to the below but pointing to different tables.
![2023-04-11_18-05-51](https://user-images.githubusercontent.com/12111398/231298176-28578527-e907-4126-a224-f964fb674d9d.png)

When I click on an event the below error pops up.
![2023-04-11_17-52-57](https://user-images.githubusercontent.com/12111398/231297889-fe65658e-cbb7-4057-b7b3-54437eabe6d7.png)

The Archive database is now showing the bar graphs.
![2023-04-11_18-00-01](https://user-images.githubusercontent.com/12111398/231297703-b15fb481-c8e8-4747-9dc8-73cdbcc69239.png)

Log: [error.log](https://github.com/NathanGibbs3/BASE/files/11205330/error.log)

**Reply by NathanGibbs3** [reply]
I just pushed to devel again, CI checks passed.
I know what the bug is, and generally where it is hiding.
Opened Issue #178 to track it.
I doubt we have seen the last of it, but we will flush it out.
Let me know how this commit works.
Fingers crossed. 🤞 😄

**Reply by mesteele** [reply]
It appears that the link issue is resolved. I will need to do more testing to see if there are any other issues as I drill down further into the options.

**Reply by NathanGibbs3** [reply]
That is excellent news. 😄 
Thank you.

**Comment by NathanGibbs3** [comment]
Thank you @mesteele for putting it all in one place.
This will make things so much easier. 👍🙂

**Comment by mesteele** [comment]
Clicking on any alert, I click the Next button and it pulls up the next alert and I chick on Next again and I get:

PHP Warning: Trying to access array offset on value of type null in D:\winids\inetpub\wwwroot\base\includes\base_state_query.inc.php on line 103
PHP Warning: Trying to access array offset on value of type null in D:\winids\inetpub\wwwroot\base\includes\base_state_query.inc.php on line 103

The above is constant and it is happening in all the categories on the second click of Next.

It appears that all the options that have a next button is doing the same or similar thing.

**Reply by NathanGibbs3** [reply]
Yep, we've got a live one!
The code there is making assumptions about a data structure type.
I'll open an issue to track it once I understand more of what is going on.
Is BASE crashing completely, or just running with the errors being displayed?
Definitely throw me an error log, so I can chase it down. :smile:

**Reply by mesteele** [reply]
I just went to my MySQL Apache2 install and it is not doing this. I'm going to do a fresh install with PostgreSQL IIS and see if it returns.

**Reply by NathanGibbs3** [reply]
You are definitely on to something. 👍 
What you found are bug tracks. 🪰 
We just need to know how BASE running in production on PHP8x is tripping it.
When you see that bug, send me the error log, so I can kill it. 🪰 ⚡ 😄

**Reply by mesteele** [reply]
BASE works as intended with Apache2 and MySQL.

BASE is getting an error. After clicking on an event for viewing anywhere in the list, then clicking Next or Previous the second time gets the error. Using W10, IIS and PostgreSQL.

Base Exits:
![2023-04-22_15-08-17](https://user-images.githubusercontent.com/12111398/233802329-773279ea-5cf2-4bc0-8de8-12839df03777.png)


Attached IIS log file:
[u_ex230422.log](https://github.com/NathanGibbs3/BASE/files/11302111/u_ex230422.log)

**Reply by NathanGibbs3** [reply]
OK, thanks.
This is definitely a deal breaker on IIS. It will get a stability tag in the issue tracker for sure.
The IIS error logs only tell me that it is crashing. The Apache error logs usually contain a PHP back trace which gives me a better idea of why it crashed.

I am chasing this down from my end as well. It does exist on PHP 7x, it just doesn't break BASE. 
Try it again on Apache and send the error log, even if it doesn't appear to break BASE, PHP 8x should complain in the log which will be helpful.

You have definitely found something that needs to be fixed. 🪰 
I'm still assessing it, then need to contain & eliminate it. ⚡ 😄

**Reply by NathanGibbs3** [reply]
Got him nicely contained in #185 & #186

**Comment by mesteele** [comment]
Here is the Apache2 error log. It appears to be getting the error but works through it.
[error.log](https://github.com/NathanGibbs3/BASE/files/11302479/error.log)

**Reply by NathanGibbs3** [reply]
Thanks. :smile:

**Reply by NathanGibbs3** [reply]
# When the returns from the class under test in various error states are all over the map.
## **Why BASE, Whyyyyy.** 💩 😞 
![Screenshot 2023-04-25 15 38 38](https://user-images.githubusercontent.com/9939563/234389026-42849949-9cfd-4f0f-b7f0-0863725bcb24.png)

**Reply by NathanGibbs3** [reply]
Fix is in the devel branch.

**Comment by mesteele** [comment]
It appears that the Next and Previous is fixed :)

However :) If $Use_Auth_System = 1; then under IIS I get a server error when I go try to access BASE and does not get to the user authentication page. Under Apache I do get an error in the log, and BASE does go to the user authentication page.

The attached log file is after loading BASE and BEFORE I enter my user credentials.
 
[error.log](https://github.com/NathanGibbs3/BASE/files/11334193/error.log)

**Reply by NathanGibbs3** [reply]
The Apache 'error' tells us that the authentication system is working as designed, at least on Apache.
Using `index.php` as the default landing page of a BASE install should not trip the authentication system, so I'm still at a loss about where the message is actually coming from.
IIS only seems to tell us 'It's broken!', well, thanks IIS, so is Chernobyl, doesn't help us fix it.

This isn't the first time the Auth system has tripped up IIS, and we still have next to no data to work with, except, 'It's broken!'.
I have a theory, of course, I'm shooting in the dark, but it 'may' help on IIS.  The theory's implementation will be out shortly.

**Reply by NathanGibbs3** [reply]
It's live in the devel branch, see what it does.

**Comment by mesteele** [comment]
I'm accessing using index.php

It is still has the same issue:

I am attaching two sets of logs for IIS and Apache. I started with new log files for each. I started the webserver, started BASE and stopped the webserver.

Apache2 logs 0 is no authentication and 1 is with authentication.
[Apache2 With 0.log](https://github.com/NathanGibbs3/BASE/files/11337022/Apache2.With.0.log)
[Apache2 With 1.log](https://github.com/NathanGibbs3/BASE/files/11337026/Apache2.With.1.log)

IIS logs 0 is no authentication and 1 is with authentication.
[IIs With 0.log](https://github.com/NathanGibbs3/BASE/files/11337056/IIs.With.0.log)
[IIs With 1.log](https://github.com/NathanGibbs3/BASE/files/11337058/IIs.With.1.log)

**Reply by NathanGibbs3** [reply]
Thanks. :smile:

**Reply by NathanGibbs3** [reply]
On my end under Linux & Apache, we have no calls to the Auth system until after credentials have been submitted.
These calls are made by `base_main.php` after `index.php` has performed an HTTP redirect to `base_main.php`.
It makes no sense if you are not even authenticating to BASE, just shutting the web server down, that the auth system is being called.

Obviously, your logs demonstrate that there is a problem.
We know where it is, even down to the routines being used, we just don't know what it is or why it is.

Wild thought, does the latest code from the devel branch work perfectly fine on both Apache & IIS running PHP 7x?

**Reply by mesteele** [reply]
I am starting the server, going to the URL index.php and it gets an Error 500. It should be stopping at the authentication prompt. 

On IIS it does the same thing with the same entry in the log file when the authentication system is set to 1 using PHP7

Not sure about Apache but I'm guessing it is the same. It will an hour or so to test that one.

**Reply by NathanGibbs3** [reply]
OK, thank you, let me know how it turns out. :smile:

**Reply by mesteele** [reply]
I have set the authentication to 1 in base using Apache2 and PHP8. I go to the index.php page and it throws the same error (not on the screen but in the log file) and it does go to the authentication page. Unlike IIS it goes to a server error page throwing an error 500 in the IIS log.

Setting the authentication back to 0 and I'm able to access BASE console with no issues.

Log file from Apache2 and PHP8:
[error.log](https://github.com/NathanGibbs3/BASE/files/11359532/error.log)

**Reply by NathanGibbs3** [reply]
I "think"🤞I found the culprit, it wasn't where we thought it was. I'll let you know when new code is pushed to devel.
Probably later today. :smile:

**Reply by NathanGibbs3** [reply]
Let me know what the latest commit to devel does.

**Reply by mesteele** [reply]
Can you verify the validity of the base_conf.php.dist file as it looks to be odd?

**Reply by NathanGibbs3** [reply]
> Let me know what the latest commit to devel does.

Actually, don't waste your time with that commit, it's broken. 💩 
I'll get a new one out later today.

**Reply by NathanGibbs3** [reply]
The `base_conf.php.dist` from the last commit is valid for anything using a codebase from commit abaf7c218ae903267477b9f7e32a081d8af949fc forward.
See Issue #66

**Reply by NathanGibbs3** [reply]
Just pushed to devel, let me know how it works.

**Reply by mesteele** [reply]
Installed into IIS and it instantly gets a server 500 error. Went back to previous commit and it works.

Installed into Apache. I started Apache, went to index.php, shut down Apache and pulled the log.

Unlike IIS, Apache did load the main page but worked through the error.

I tried PHP7 and PHP8 and both did the same thing.

Apache2 Log: 
[error.log](https://github.com/NathanGibbs3/BASE/files/11376348/error.log)

I stopped Apache2, deleted the error.log, set Auth_System to 1, started Apache2, went to index.php and I instantly got a 500 error. I stopped Apache2 and I'm attaching that error.log file

Apache2 Log:
[error.log](https://github.com/NathanGibbs3/BASE/files/11376492/error.log)

**Reply by NathanGibbs3** [reply]
> Installed into IIS and it instantly gets a server 500 error. Went back to previous commit and it works.
> 

Wait, which previous commit?! We've been talking about BASE NOT working on IIS for the last week!

> Installed into Apache. I started Apache, went to index.php, shut down Apache and pulled the log.
> 
> Unlike IIS, Apache did load the main page but worked through the error.
> 

Is it working or not working on IIS? 😕 

> I tried PHP7 and PHP8 and both did the same thing.
> 
> Apache2 Log: [error.log](https://github.com/NathanGibbs3/BASE/files/11376348/error.log)
> 

This log file tells me that BASE found it's config file and started.
Also, that it is at least last night's commit to devel, ( lets hope not ), or this mornings commit, ( lets hope so).

> I stopped Apache2, deleted the error.log, set Auth_System to 1, started Apache2, went to index.php and I instantly got a 500 error. I stopped Apache2 and I'm attaching that error.log file
> 
> Apache2 Log: [error.log](https://github.com/NathanGibbs3/BASE/files/11376492/error.log)

Are you positive that you are not running last night's commit?
I had the same issue with last night's commit, which is why I fixed it and pushed a new commit to devel earlier today.

**Reply by mesteele** [reply]
I just pulled the latest devl and installed just to make sure and loading index.php in IIS gets a 500 error.

I installed the latest devl into Apache.

Log 1: This came from the last devel that I installed about a week ago. All I did was stop the server, delete the error.log, started the server, went to index.php, stopped the server, and collected the log below.
[log-1.log](https://github.com/NathanGibbs3/BASE/files/11377587/log-1.log)

Log 2: This came from the **latest** devel. All I did was stop the server, delete the error.log, started the server, went to index.php, stopped the server, and collected the log below.
[log-2.log](https://github.com/NathanGibbs3/BASE/files/11377701/log-2.log)

Log 3: This came from the **lastest** devel. All I did change the Use_Auth_System to 1, stop the server, deleted the error.log, started the server, went to index.php, stopped the server, and collected the log below.
[log-3.log](https://github.com/NathanGibbs3/BASE/files/11377729/log-3.log)

**Reply by NathanGibbs3** [reply]
OK, thanks.
I'm curious about this.
Have we been having issues with the Auth system breaking on IIS for about a month?

I "think" I know whats happening, why, and which commit started it.
Of course, I've said that before, lol. :-)

**Reply by NathanGibbs3** [reply]
OK, we have Issue #189 to track this one, and a new commit to devel that "hopefully" fixes it.
Let me know. :smile:

**Reply by mesteele** [reply]
Not sure for how long. It works in the last BASE I used that was specifically for PHP 7.

**Reply by NathanGibbs3** [reply]
I think I made the change responsible for this in early April when we were working on #161
Hopefully what is in devel now, corrects that. :smile:

**Reply by mesteele** [reply]
Just downloaded and same issue on IIS, server 500.

I go to Apache and with Auth set to 0 and the log files are the same as the above Log 1 and log 2 files

When IIS does go to the main base page the Apache log file is Log 1.

When I set the auth to 1, BASE now goes to the login screeen and I'm able to log in.

A lot looks good but for just limited testing I found that when I go to Cache & Status I get the below:
![2023-05-03_8-10-14](https://user-images.githubusercontent.com/12111398/235913271-9a859c96-9d9f-4868-9c44-bc8b25e12340.png)

**Reply by NathanGibbs3** [reply]
> Just downloaded and same issue on IIS, server 500.
> 

So, BASE on IIS is still broken apparently.

> I go to Apache and with Auth set to 0 and the log files are the same as the above Log 1 and log 2 files
> 

Log-1 & log-2 from above only indicated normal runs of BASE with no issues, so nothing broke on Apache with the Auth system off.
:smile: 

> When IIS does go to the main base page the Apache log file is Log 1.
> 

First, the Apacha log file wouldn't be relevant for an IIS run.
So BASE is not longer broken on IIS, appearantly.

Hmm, Schrodinger's IIS or Schrodinger's BASE,
It both runs and doesn't run. 😕 

> When I set the auth to 1, BASE now goes to the login screeen and I'm able to log in.
> 

Which Web Server are we referring to here?
Obviously, we fixed something, somewhere?

> A lot looks good but for just limited testing I found that when I go to Cache & Status I get the below: ![2023-05-03_8-10-14](https://user-images.githubusercontent.com/12111398/235913271-9a859c96-9d9f-4868-9c44-bc8b25e12340.png)

Which web server did that?
If it's Apache send over the logs.

**Reply by mesteele** [reply]
I can only test on Apache2 as IIS instantly goes to a server 500 error with the latest commit. The IIS logs are useless because it only shows a 500 error. Any log files i sent are from Apache.

I'm using two different version of BASE.

What is now fixed is on Apache the Auth set to 1 now works but I'm unable to test on IIS because it still crashes accessing the index.php file.

I ran both versions on Apache2 and it produced the two log files.

Log file 1 is from the last commit that BASE would open, and not get the error 500 when accessing index.php on IIS.

Log file 2 is from the latest commit that gets an error 500 on IIS when accessing index.php on IIS.

Both versions run on Apache2. Log file 2 appears to show an error with the conf file, where log file 1 does not.

I can remove the conf file in the latest commit and IIS does the same thing, an error 500. It's not even recognizing the missing conf file on IIS..

I would think the output of log 1 was normal?

The above may not be causing the IIS issue?

What is now fixed is that Auth set to 1 now gets to the login page and I can login on Apache2. I'm unable to test anything in IIS as it crashes trying to load the index.php page.

**Reply by NathanGibbs3** [reply]
> I can only test on Apache2 as IIS instantly goes to a server 500 error with the latest commit. The IIS logs are useless because it only shows a 500 error. Any log files i sent are from Apache.
> 

OK, so at the moment, BASE dies on IIS all the time, period.
With auth system off, index.php succcessfully redirects the user to base_main.php, where BASE dies for yet unexplained reasons.
With auth system on, BASE dies for yet unexplained reasons on index.php

I know very little about being a sysadmin for a WAMP/WIMP stack.
This one is yours. :smile:
- [ ] Figure out how to get useful error logs out of IIS running a PHP application.

Maybe the `error_log` item in the PHP configuration file has something to do with it.
Google might have an answer too, 
I doubt you are the first to try to debug a PHP app running on IIS.
If you can figure that out, and get useful logs, I'll have a chance of figuring out why BASE is suddenly dying on IIS.

> I'm using two different version of BASE.
> 
> What is now fixed is on Apache the Auth set to 1 now works but I'm unable to test on IIS because it still crashes accessing the index.php file.
> 
> I ran both versions on Apache2 and it produced the two log files.
> 
> Log file 1 is from the last commit that BASE would open, and not get the error 500 when accessing index.php on IIS.
> 

OK, so something between that commit and now broke BASE on IIS.
We need two things.
- [ ] The git hash of the commit that works on IIS. ( the latest commit that does work on IIS )
- [ ] The git hash of the first commit that doesn't work on IIS. ( the earliest descendant commit from above that does not work on IIS )

The IIS breaking bug will be between those two somewhere.

> Log file 2 is from the latest commit that gets an error 500 on IIS when accessing index.php on IIS.
> 
> Both versions run on Apache2. Log file 2 appears to show an error with the conf file, where log file 1 does not.
> 

log-1 was generated by a BASE before the BASE kernel was implemented.
log-2 was generated by a BASE in which the kernel is implemented.

For web apps, the error_log is basically, the same concept as syslog for Linux or the event log for windows.

The "errors" in log-2 are just the BASE kernel logging events.
log-2 doesn't contain anything that indicates an issue.

> I can remove the conf file in the latest commit and IIS does the same thing, an error 500. It's not even recognizing the missing conf file on IIS..
> 
> I would think the output of log 1 was normal?
> 
> The above may not be causing the IIS issue?
> 
> What is now fixed is that Auth set to 1 now gets to the login page and I can login on Apache2. I'm unable to test anything in IIS as it crashes trying to load the index.php page.

**Reply by mesteele** [reply]
> OK, so at the moment, BASE dies on IIS all the time, period.

Absolutely, IIS dies all the time when accessing index.php, and it doesn't matter how the Auth system is configured.

It's like it hits a wall when the index.php file is accessed.

> With auth system off, index.php succcessfully redirects the user to base_main.php, where BASE dies for yet unexplained reasons.

Not sure it just dies hitting the index.php page and never gets to the base_main.php page.

> With auth system on, BASE dies for yet unexplained reasons on index.php

Absolutly.

I'll look into getting logs from PHP.

**Reply by mesteele** [reply]
Ok, so her is a weird one. It now works and here is the weird part:

error_reporting = E_ALL & ~E_DEPRECATED & ~E_STRICT
error_log = d:\winids\php\logs\php_errors.log

Once I activated error checking above it worked. It appears error checking needs to be turned on for some reason.

With Auth set to 0 I now get to BASE. I haven't done much testing there, but it all looks good.

With Auth set to one I get an error 500 screen. I'm attaching the log file for IIS and PHP.
[php_errors.log](https://github.com/NathanGibbs3/BASE/files/11391355/php_errors.log)

**Reply by NathanGibbs3** [reply]
> Ok, so her is a weird one. It now works and here is the weird part:
> 
> error_reporting = E_ALL & ~E_DEPRECATED & ~E_STRICT error_log = d:\winids\php\logs\php_errors.log
> 
> Once I activated error checking above it worked. It appears error checking needs to be turned on for some reason.
> 

Awesome, that's great news. 👍 😄 

> With Auth set to 0 I now get to BASE. I haven't done much testing there, but it all looks good.
> 
> With Auth set to one I get an error 500 screen. I'm attaching the log file for IIS and PHP. [php_errors.log](https://github.com/NathanGibbs3/BASE/files/11391355/php_errors.log)

Looks like the same stack trace we have in log-3 from Apache
Try this on both Apache & IIS..

From your browser's address bar got to
`http://winids/base_denied.php`

Maybe we can narrow this down to a specific web server.

**Reply by mesteele** [reply]
On IIS. with Auth set to 0 it goes to:
![2023-05-04_9-45-25](https://user-images.githubusercontent.com/12111398/236225141-48192824-38b5-4082-b59d-8e8120e7a3ab.png)
Log:
[auth-0 php_errors.log](https://github.com/NathanGibbs3/BASE/files/11398114/auth-0.php_errors.log)

On IIS. with Auth set to 1 it goes to:
![2023-05-04_9-48-54](https://user-images.githubusercontent.com/12111398/236226647-798f77cf-bd2d-4708-be25-c4631bc237a7.png)
Log:
[auth-1 php_errors.log](https://github.com/NathanGibbs3/BASE/files/11398143/auth-1.php_errors.log)

**Reply by NathanGibbs3** [reply]
Does it work OK on Apache?

I think we have a case of IIS not supplying some standard PHP global server vars.
Putting the below code into a file in your IIS site, then pointing your browser to the file should dump out what IIS actually does supply in the `$_SERVER` PHP global var.

```php
<?php
print '<html><head></head><body>';
var_dump($_SERVER);
print '</body></html>';
?>
```

BASE makes some code execution decisions based on `$_SERVER['SCRIPT_NAME']`, if IIS is not supplying that, then we know what is behind this.

Knowing what IIS actually does supply, will give me a clue on how to code around this MicroSuck induced breakage. 🧯

**Reply by mesteele** [reply]
Not by Apache right now. Attached is the screen dump.
[print.txt](https://github.com/NathanGibbs3/BASE/files/11399183/print.txt)

**Reply by mesteele** [reply]
Used this to print out:

```
<?php
    ob_start();
    var_dump($_SERVER) ;
    $dump = ob_get_contents();
    ob_end_clean();

    echo "<pre> $dump </pre>";
?>
```

Log:
[print.txt](https://github.com/NathanGibbs3/BASE/files/11399405/print.txt)

**Reply by NathanGibbs3** [reply]
OK, thanks, this is helping

In your conf file, what is `$BASE_urlpath` set to?

**Reply by mesteele** [reply]
$BASE_urlpath = 'http://winids';

**Reply by NathanGibbs3** [reply]
Seriously!!!!!! 🤯 

This is from the BASE conf file itself.
```
 Set the base_urlpath to the url location that is the root of your BASE install.
 This must be set for BASE to function! Do not include a trailing slash!
 But also put the preceding slash. e.g. Your URL is http://127.0.0.1/base
 set this to /base
```

I could be wrong, but I 'think', we actually found the bug this time. :smile:

The URL you use is `http://winids`
So `$BASE_urlpath` should be set to an empty string just `''`

So, I've spent weeks chasing around BASE looking for a bug that is not there.

**Reply by mesteele** [reply]
Not sure but there has been thousands of installs and no problem like this has been reported. For some reason PHP8 is noticing it, where PHP 5-7 didn't care.

It appears to be working correctly.

**Reply by NathanGibbs3** [reply]
> On IIS. with Auth set to 0 it goes to: ![2023-05-04_9-45-25](https://user-images.githubusercontent.com/12111398/236225141-48192824-38b5-4082-b59d-8e8120e7a3ab.png) Log: [auth-0 php_errors.log](https://github.com/NathanGibbs3/BASE/files/11398114/auth-0.php_errors.log)

Then there has been thousands of installs exposing admin functions and version information to unauthenticated users due to system misconfiguration.

The underlying disclosure issues were addressed in October of 2020 with fixes for Issues #78 & #89 
The next BASE Kernel will keep this misconfiguration scenario from re-exposing these issues.

**Reply by mesteele** [reply]
Thank you for all your hard work, it is appreciated :)

**Comment by mesteele** [comment]
I'm in the main snort database, I go into Cache and Status. and want to clear all the alerts. I try but the screen just blinks and nothing gets deleted.
![2023-05-03_19-34-26](https://user-images.githubusercontent.com/12111398/236076370-44c91157-f696-4494-ac5e-48fc177d9814.png)

 The checkmark is on the archive database so I switch to the archive database and try to copy or move events and both give this:
![2023-05-03_19-41-05](https://user-images.githubusercontent.com/12111398/236076409-d5687704-1974-4b4d-a837-bc8cbb610cee.png)

None of the functions seems to be working in Cache and Status in either MySQL or PostgreSQL.

**Reply by NathanGibbs3** [reply]
That indicates that the Archive DB flag is enabled in conf.
It does not guarantee the existence, structure of, or access to an archive DB.
See Issue #102 for more information. 
It appears that you use the archive DB functionality, maybe you could point us to information that would be helpful in closing #102. 🙂

**Reply by mesteele** [reply]
I'm not sure what you are needing?

I always manually create both databases. Go [here](http://www.winsnort.com/tutorials/article/2-installing-an-iis-web-server-logging-events-to-a-mysql-database/) and there is a section called Installing the MySQL Database Server on how I Install MySQL. Then go to the section Configuring the MySQL Database Server to see how I manually configure the databases. Then go to the section titled Configuring BASE the Windows Intrusion Detection Systems (WinIDS) Security Console to see how I configured BASE to access both databases.

**Reply by NathanGibbs3** [reply]
Thank you, I've never run BASE with an archive DB, as there was no documentation in BASE on how to set one up. 
So it appears that the archive DB has the same structure as the main snort/ACID/BASE alert DB.
Couple of questions.
1. Is Barnyard being actively maintained?
   - What I have for them is https://github.com/firnsy/barnyard2
   - As barnyard, like BASE is under GPL, I'm thinking about incorporating their DB creation code into a future version of BASE,. That way BASE could bootstrap it's own required DB infrastructure during setup, if needed.
   - Obviously, it would have to check for a pre existing DB structure. We wouldn't want an upgrade trashing a production DB. 🙀 🤯 
2. What are your thoughts on removing the authentication tables from the archive DB is a future BASE? 
    - It seems like a bad idea to have login credentials laying around in the archive DB.
 
Your thoughts on any of this would be appreciated. :smile:

**Reply by mesteele** [reply]
>So it appears that the archive DB has the same structure as the main snort/ACID/BASE alert DB.

Yes, you could just change event database name to archive and it would done. I have no idea how it works as it has been years since I tried it. I don't think it does PostgreSQL.

>Couple of questions.

>Is Barnyard being actively maintained?

No, it has not been maintained for a few years now.

>What I have for them is https://github.com/firnsy/barnyard2

Yes, and I do the compiling for Windows.

> As barnyard, like BASE is under GPL, I'm thinking about incorporating their DB creation code into a future version of BASE,. 
That way BASE could bootstrap it's own required DB infrastructure during setup, if needed.

The only thing Barnyard2 does is read in the snort event log, decode it and move the decoded events to whatever database Barnyard2 is configured to use.

> Obviously, it would have to check for a pre existing DB structure. We wouldn't want an upgrade trashing a production DB. 🙀 🤯

See the above.

>What are your thoughts on removing the authentication tables from the archive DB is a future BASE?

That would be a huge amoubnt of work. For the obvious, it would be better to move all that to a Users Group in the OS.

>It seems like a bad idea to have login credentials laying around in the archive DB.

It's never a good idea but I think the original programmer was trying to make ACID as all inclusive as it could.

**Comment by mesteele** [comment]
PHP apparently has two active versions being developed (8.1.xx and 8.2.xx).

I am using PHP 8.2.5 to test on, so should I move back to 8.1.xx so as not to leapfrog over the version you are using?

> What parts of BASE I have built unit tests for get tested on travis-ci from PHP 5.2 - 8-1.

**Reply by NathanGibbs3** [reply]
Run it on whichever PHP you want to.
BASE was originally designed to run on 4.0.4+. I've tried to keep it that way. As mentioned, I test it on the PHP major.minor combos offered by Travis-ci. That way, as things get added or changed, we don't negate the viability of BASE on older systems.
